src/btree/bt_curnext.c

```diff
@@ -431,6 +431,147 @@ __wt_btcur_iterate_setup(WT_CURSOR_BTREE *cbt)
+	default:
+		break;
+	}
+	return ("UNKNOWN last op");
+}
+
+/*
+ * __wt_set_last_op --
+ *	Set the last operation.
+ */
+void
+__wt_set_last_op(WT_CURSOR_BTREE *cbt, int v)
+{
+	int i;
+
+	for (i = 20; --i > 0;)
+		cbt->last_op[i] = cbt->last_op[i - 1];
+	cbt->last_op[0] = v;
+}
+
+/*
+ * __wt_track_location_name --
+ *	Return the last location name.
+ */
+static const char *
+__wt_track_location_name(int v)
+{
+	switch (v) {
+	case WT_TRACKOP_CUR_NEXT: return ("cursor-next");
+	case WT_TRACKOP_ROW_NEXT_START: return ("row-next-start");
+	case WT_TRACKOP_ROW_NEXT_STOP: return ("row-next-stop");
+	case WT_TRACKOP_WALK_START: return ("walk-start");
+	case WT_TRACKOP_WALK_STOP: return ("walk-stop");
+	case WT_TRACKOP_CUR_RETURN: return ("cursor-next-return");
+	default:
+		break;
+	}
+	return ("UNKNOWN location");
+}
+
+/* 
+ * __wt_key_order_check --
+ *	Check key ordering for cursor movements.
+ */
+static void
+__wt_key_order_check(WT_SESSION_IMPL *session, WT_CURSOR_BTREE *cbt)
+{
+	WT_BTREE *btree;
+	WT_DECL_ITEM(a);
+	WT_DECL_ITEM(b);
+	WT_ITEM *key;
+	WT_DECL_RET;
+	int cmp, i;
+
+	btree = S2BT(session);
+	key = &cbt->iface.key;
+
+	if ((ret = __wt_compare(
+	    session, btree->collator, cbt->lastkey, key, &cmp)) != 0)
+		WT_ERR_MSG(session, ret,
+		    "SUPPORT-1531: comparison function failed");
+	if (cmp < 0)
+		return;
+
+	WT_ERR(__wt_scr_alloc(session, 512, &a));
+	WT_ERR(__wt_scr_alloc(session, 512, &b));
+
+	WT_ERR(__wt_buf_set_printable(session, a, key->data, key->size));
+	WT_ERR(__wt_buf_set_printable(
+	    session, b, cbt->lastkey->data, cbt->lastkey->size));
+
+	__wt_errx(session,
+	    "SUPPORT-1531: WT_CURSOR.next out-of-order keys: returning %.*s "
+	    "which is not larger than a previously returned key %.*s",
+	    (int)a->size, (const char *)a->data,
+	    (int)b->size, (const char *)b->data);
+
+	__wt_errx(session,
+	    "SUPPORT-1531: WT_CURSOR last operations (most-recent first): "
+	    "%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, "
+	    "%s, %s, %s, %s, %s, %s, %s, %s, %s, %s",
+	    __wt_last_op_name(cbt->last_op[0]),
+	    __wt_last_op_name(cbt->last_op[1]),
+	    __wt_last_op_name(cbt->last_op[2]),
+	    __wt_last_op_name(cbt->last_op[3]),
+	    __wt_last_op_name(cbt->last_op[4]),
+	    __wt_last_op_name(cbt->last_op[5]),
+	    __wt_last_op_name(cbt->last_op[6]),
+	    __wt_last_op_name(cbt->last_op[7]),
+	    __wt_last_op_name(cbt->last_op[8]),
+	    __wt_last_op_name(cbt->last_op[9]),
+	    __wt_last_op_name(cbt->last_op[10]),
+	    __wt_last_op_name(cbt->last_op[11]),
+	    __wt_last_op_name(cbt->last_op[12]),
+	    __wt_last_op_name(cbt->last_op[13]),
+	    __wt_last_op_name(cbt->last_op[14]),
+	    __wt_last_op_name(cbt->last_op[15]),
+	    __wt_last_op_name(cbt->last_op[16]),
+	    __wt_last_op_name(cbt->last_op[17]),
+	    __wt_last_op_name(cbt->last_op[18]),
+	    __wt_last_op_name(cbt->last_op[19]));
+
+	for (i = cbt->next_track - 1; i >= 0; --i)
+		__wt_errx(session,
```

Comment:
Assuming that cursors are calloc'ed (or otherwize zero initialized) does it make sense to skip printing in these loops if .locate == 0? Or make __wt_track_location_name have a special case for 0?